runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceRDD.java

```diff
@@ -0,0 +1,199 @@
+      // when running on YARN/SparkDeploy it's the result of max(totalCores, 2).
+      // when running on Mesos it's 8.
+      // when running local it's the total number of cores (local = 1, local[N] = N,
+      // local[*] = estimation of the machine's cores).
+      // ** the configuration "spark.default.parallelism" takes precedence over all of the above **
+      this.numPartitions = sc.defaultParallelism();
+    }
+
+    private static final long DEFAULT_BUNDLE_SIZE = 64 * 1024 * 1024;
+
+    @Override
+    public Partition[] getPartitions() {
+      long desiredSizeBytes = DEFAULT_BUNDLE_SIZE;
+      try {
+        desiredSizeBytes = source.getEstimatedSizeBytes(
+            runtimeContext.getPipelineOptions()) / numPartitions;
+      } catch (Exception e) {
+        LOG.warn("Failed to get estimated size of bundle, default is {} bytes.",
+            DEFAULT_BUNDLE_SIZE);
+      }
+      try {
+        List<? extends Source<T>> partitionedSources = source.splitIntoBundles(desiredSizeBytes,
+            runtimeContext.getPipelineOptions());
+        Partition[] partitions = new SourcePartition[partitionedSources.size()];
+        for (int i = 0; i < partitionedSources.size(); i++) {
+          partitions[i] = new SourcePartition<>(id(), i, partitionedSources.get(i));
+        }
+        return partitions;
+      } catch (Exception e) {
+        throw new RuntimeException("Failed to create partitions.", e);
+      }
+    }
+
+    @Override
+    public scala.collection.Iterator<WindowedValue<T>> compute(final Partition split,
+                                                               TaskContext context) {
+      final Iterator<WindowedValue<T>> iter = new Iterator<WindowedValue<T>>() {
+        @SuppressWarnings("unchecked")
+        SourcePartition<T> partition = (SourcePartition<T>) split;
+        BoundedSource.BoundedReader<T> reader = createReader(partition);
+
+        private boolean finished = false;
+        private boolean started = false;
+        private boolean closed = false;
+
+        @Override
+        public boolean hasNext() {
+          try {
+            if (!started) {
+              started = true;
+              finished = !reader.start();
+            } else {
+              finished = !reader.advance();
+            }
+            if (finished) {
+              // safely close the reader if there are no more elements left to read.
+              closeIfNotClosed();
+            }
+            return !finished;
+          } catch (IOException e) {
+            throw new RuntimeException("Failed to read from reader.", e);
+          }
+        }
+
+        @Override
+        public WindowedValue<T> next() {
+          return WindowedValue.of(reader.getCurrent(), reader.getCurrentTimestamp(),
+              GlobalWindow.INSTANCE, PaneInfo.NO_FIRING);
+        }
+
+        @Override
+        public void remove() {
+          // do nothing.
```

Comment:
Is the runner expected to call this? If not, can probably throw an `UnsupportedOperationException` just to be more explicit about what the iterator supports, and/or to see when we're interacted with unexpectedly.